WIP - Stream results from the database to JSON

senior opened the pull request:
This commit uses reducible query results from clojure.java.jdbc to
stream application database query results from to the response as
JSON. Previously we streamed the JSON serialization part, but not the
query portion.

This commit is more intended as an RFC and only switches over the
`/database/:id/fields` endpoint to use this. Testing on my machine, on
large synthetic schema, shows it uses about half the memory that the
previous code used and executes in roughly the same wall clock time.

This commit relies on related query-reducible functionality in Toucan
that hasn't been merged yet, hence the dependency on Toucan
1.1.4-SNAPSHOT.

iethree committed:
Stream results from the database to JSON

This commit uses reducible query results from clojure.java.jdbc to
stream application database query results from to the response as
JSON. Previously we streamed the JSON serialization part, but not the
query portion.

This commit is more intended as an RFC and only switches over the
`/database/:id/fields` endpoint to use this. Testing on my machine, on
large synthetic schema, shows it uses about half the memory that the
previous code used and executes in roughly the same wall clock time.

This commit relies on related query-reducible functionality in Toucan
that hasn't been merged yet, hence the dependency on Toucan
1.1.4-SNAPSHOT.

sbelak commented on src/metabase/api/database.clj:
```diff
@@ -261,22 +270,116 @@
+        ;; Hydrate in 10,000 row batches
+        (map #(hydrate % :table))
+        ;; This will flatten the chunks about out
+        cat
```
Any reason why not mapcat?

senior replied:
`mapcat` is better here 👍

sbelak commented on src/metabase/api/database.clj:
```diff
@@ -261,22 +270,116 @@
+               ;; This eduction will help us know that we've successfully executed a query and recevied a result from
+               ;; the result set. Up to this point, we have established a connection to the database but not run the
+               ;; query. We'll check for the first row first before running the `transducer` on the results
+               (eduction (map (call-on-first-row on-first-row)) transducer reducible-q))
```
Why do you need eduction here? If I follow, the reducible is immediately (and only once) consumed by run! Can't you simply do:
```
(transduce (comp (map (call-on-first-row on-first-row)) transducer) reducible-q)
```

senior replied:
Seems like an equal trade there. I guess the benefits of the above approach is that it doesn't wrap the reducible in the defrecord. I'd be curious what you think about this being nested in a with-open vs. the side-affecty transducer handling the opening/closing of the writer.

sbelak replied:
If all things were equal, I'd stuff all state in the transducer, however your current approach is more robust in how it handles exceptions. If all state handling is in the transducer and there's an exception in the xf, you're left dangling.

sbelak commented on src/metabase/api/database.clj:
```diff
@@ -261,22 +270,116 @@
+    ;; Block until a row or an exception has been delivered to `p`. This will catch something like SQL query that
+    ;; fails to compile or references something that doesn't exist and allow that exception to be returned as if the
+    ;; query was running in the HTTP thread.
+    (throw-if-exception p)
```
This is going to be super subjective: but this approach with promises and cross-thread communication is making my brain hurt. Couldn't this be done with something state-machine like?

iethree committed:
Add `defendpoint-streaming` that will stream from db to response

The pieces to implement streaming are the same as the previous commit,
just packaged in a new defendpoint macro.

senior commented:
Added a new `defendpoint-streaming` that packages up the previous code into a better api. I think `defendpoint` could use a refactor too, but this doesn't feel like the right time to do that.

camsaul commented:
Closing since @senior is not longer actively working on Metabase 😢 💔
